File: packages/typescript/src/adapters/credentials.ts
import { execFile } from "node:child_process";
import { promises as fs } from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { promisify } from "node:util";
import { ProfileSetupFailedError } from "../errors.js";

const execFileP = promisify(execFile);

export async function readClaudeCredentials(): Promise<string> {
  if (process.platform === "darwin") {
    try {
      const { stdout } = await execFileP("security", [
        "find-generic-password",
        "-s",
        "Claude Code-credentials",
        "-a",
        process.env["USER"] ?? "",
        "-w",
      ], { timeout: 10_000 });
      const trimmed = stdout.trim();
      if (trimmed.length > 0) {
        return trimmed.endsWith("\n") ? trimmed : trimmed + "\n";
      }
    } catch {
      // fall through
    }
  }
  const fallback = path.join(os.homedir(), ".claude", ".credentials.json");
  try {
    return await fs.readFile(fallback, "utf-8");
  } catch {
    throw new ProfileSetupFailedError(
      "Claude Code credentials not found in macOS Keychain ('Claude Code-credentials') or ~/.claude/.credentials.json. Run `claude` once to log in before running.",
      { reason: "claude_credentials_missing" },
    );
  }
}

export async function readCodexAuth(): Promise<string> {
  const src = path.join(os.homedir(), ".codex", "auth.json");
  try {
    return await fs.readFile(src, "utf-8");
  } catch {
    throw new ProfileSetupFailedError(
      "~/.codex/auth.json not found. Run `codex` once to log in before running.",
      { reason: "codex_auth_missing", path: src },
    );
  }
}

File: packages/rust/src/adapters/credentials.rs
use crate::error::Error;
use std::path::PathBuf;
#[cfg(target_os = "macos")]
use std::process::Command;

pub(crate) fn read_claude_credentials() -> Result<String, Error> {
    #[cfg(target_os = "macos")]
    {
        let output = Command::new("security")
            .args([
                "find-generic-password",
                "-s",
                "Claude Code-credentials",
                "-a",
                std::env::var("USER").unwrap_or_default().as_str(),
                "-w",
            ])
            .output();
        if let Ok(out) = output {
            if out.status.success() {
                let s = String::from_utf8_lossy(&out.stdout).trim().to_string();
                if !s.is_empty() {
                    return Ok(if s.ends_with('\n') {
                        s
                    } else {
                        format!("{s}\n")
                    });
                }
            }
        }
    }
    let home = home_dir();
    let fallback = home.join(".claude").join(".credentials.json");
    std::fs::read_to_string(&fallback).map_err(|_| Error::CredentialsMissing {
        reason: "Claude Code credentials not found in macOS Keychain ('Claude Code-credentials') or ~/.claude/.credentials.json".into(),
        adapter_id: "claude-code".into(),
    })
}

pub(crate) fn read_codex_auth() -> Result<String, Error> {
    let home = home_dir();
    let src = home.join(".codex").join("auth.json");
    std::fs::read_to_string(&src).map_err(|_| Error::CredentialsMissing {
        reason: format!("{} not found. Run `codex` once to log in.", src.display()),
        adapter_id: "codex".into(),
    })
}

fn home_dir() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("/"))
}
